torch/_higher_order_ops/invoke_subgraph.py

```diff
@@ -0,0 +1,231 @@
+# mypy: allow-untyped-decorators
+# mypy: allow-untyped-defs
+
+
+import torch
+import torch.utils._pytree as pytree
+from torch._C import DispatchKey
+from torch._dispatch.python import suspend_functionalization
+from torch._higher_order_ops.utils import (
+    _from_fun,
+    _maybe_reenter_make_fx,
+    clone_outputs_aliasing_inputs,
+    get_dummy_aot_autograd_config,
+    prepare_fw_with_masks,
+    reenter_make_fx,
+)
+from torch._ops import HigherOrderOperator
+from torch._subclasses import FakeTensorMode
+from torch._subclasses.functional_tensor import disable_functional_mode
+from torch.fx.experimental.proxy_tensor import (
+    disable_proxy_modes_tracing,
+    ProxyTorchDispatchMode,
+    track_tensor_tree,
+)
+from torch.fx.graph_module import GraphModule
+
+
+invoke_subgraph_counter = 0
+
+
+class InvokeSubgraphHOP(HigherOrderOperator):
+    def __init__(self) -> None:
+        super().__init__("invoke_subgraph")
+
+    # identifier is setup by upper part of the stack. This helps us in
+    # identifying two invoke_subgraph calls have same subgraph.
+    def __call__(
+        self,
+        subgraph: GraphModule,
+        identifier: str,
+        operands,
+    ):
+        return super().__call__(subgraph, identifier, operands)
+
+
+invoke_subgraph = InvokeSubgraphHOP()
+
+
+def create_invoke_subgraph_op(subgraph, operands):
+    # Helper to set the identifier for the invoke_subgraph op. We ensure that
+    # each identifier is different here. Its torch.compile responsibility to
+    # replace identifiers for subgraphs that are identical.
+    global invoke_subgraph_counter
+    identifier = f"invoke_subgraph_{invoke_subgraph_counter}"
+    invoke_subgraph_counter += 1
+    return invoke_subgraph(subgraph, identifier, operands)
+
+
+def trace_joint_graph(fn, fw_inputs, fw_outputs):
+    """
+    Naively trace out a joint graph. This simplifies the reconstruction of joint
+    graph in the min-cut partitioner later on.
+    """
+    from torch._functorch.aot_autograd import create_joint
+
+    dummy_aot_config = get_dummy_aot_autograd_config()
+
+    def joint_fn(*primals_and_tangents):
+        primals = primals_and_tangents[: len(fw_inputs)]
+        tangents = primals_and_tangents[len(fw_inputs) :]
+
+        fw_outs, grads = create_joint(
+            prepare_fw_with_masks(fn), aot_config=dummy_aot_config
+        )(primals, tangents)
+
+        maybe_clone = clone_outputs_aliasing_inputs(primals_and_tangents)
+
+        return pytree.tree_map(maybe_clone, list(fw_outs) + grads)
+
+    primals = list(fw_inputs)
+    # This assumes that the tangent strides match fw_outputs strides. Check the
+    # InvokeSubgraphAutogradOp backward op for the contiguous call.
+    tangents = [_from_fun(out) for out in fw_outputs]
+
+    joint_operands = primals + tangents
+
+    return _maybe_reenter_make_fx(joint_fn)(*joint_operands)
+
+
+def create_fw_bw_graph(subgraph, operands):
```

Comment:
there's already one in _higher_order_ops/utils.py, can we reuse it?

Reply:
We can't. The difference is in the bwd graph which is a joint graph that returns both fwd and backward outputs, and then autograd.Function ignores the forward outputs. This is to simplify the min-cut partitioner later on.

We might want to do a refactor later on, where all HOPs use this function. But, for now, we should keep it for invoke_subgraph only.